workers/scanner.go

```diff
@@ -0,0 +1,151 @@
+package workers
+
+import (
+	"context"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/skynetlabs/pinner/database"
+	"github.com/skynetlabs/pinner/skyd"
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/NebulousLabs/threadgroup"
+	"gitlab.com/SkynetLabs/skyd/build"
+)
+
+/**
+ PHASE 1: <DONE>
+ - scan the DB once a day for underpinned files
+ - when you find an underpinned file that's not currently locked
+	- lock it
+	- pin it locally and add the current server to its list
+	- unlock it
+
+ PHASE 2:
+ - calculate server load by getting the total number and size of files pinned by each server
+ - only pin underpinned files if the current server is in the lowest 20% of servers, otherwise exit before scanning further
+
+ PHASE 3:
+ - add a second scanner which looks for skylinks which should be unpinned and unpins them from the local skyd.
+*/
+
+var (
+	// SleepBetweenPins defines how long we'll sleep between pinning files.
+	// We want to add this sleep in order to prevent a single server from
+	// grabbing all underpinned files and overloading itself. We also want to
+	// allow for some time for the newly pinned files to reach full redundancy
+	// before we pin more files.
+	SleepBetweenPins = build.Select(
+		build.Var{
+			Standard: 10 * time.Second,
+			Dev:      time.Second,
+			Testing:  time.Millisecond,
+		}).(time.Duration)
+
+	// SleepBetweenScans defines how often we'll scan the DB for underpinned
+	// skylinks.
+	SleepBetweenScans = build.Select(build.Var{
+		// In production we want to use a prime number of hours, so we can
+		// de-sync the scan and the sweeps.
+		Standard: 19 * time.Hour,
```

Comment:
REQ: I would make this shorter. If we lose a server, we want high confidence that other servers are picking up pin requests asap. 

For de-syncing, you could use 47min.

Can you elaborate on why you thought 19 hours was a good place to start? Does this operation have a very high performance cost o the mongo cluster?  it is all reads right, until we actually need to pin something, so that should just be reading locally and not locking up the master node.

Reply:
This doesn't affect Mongo much. It might affect it a bit if we start using a locking library but that still won't be enough to make us run it less often than ideal. The short sleep added after each pin makes sure that the load is spread over time and doesn't come as a single spike, so we shouldn't worry about it at all.

I chose 19 hours because it all conversations until this one, we've mentioned something along the lines of "once a day we'll scan for underpinned skylinks".

Also, and this is important, if we lose a server at the current stage of the service, *nothing will happen*. We need to call Pinner and tell it explicitly that a server was lost. This will be implemented under https://linear.app/skynetlabs/issue/SKY-297/implement-an-endpoint-which-marks-a-given-server-as-dead . There's nothing stopping us from triggering an extra sweep manually, after a server is pronounced dead. 

A good motivation for lowering this period would be if we want to use Pinner for driving extra redundancy, i.e. whenever any server pins something, we would have another server pin it as well. And we need this to happen fast. This is fine but I was under the impression that we don't want to do that at this point.

Reply:
> Also, and this is important, if we lose a server at the current stage of the service, nothing will happen. We need to call Pinner and tell it explicitly that a server was lost. This will be implemented under https://linear.app/skynetlabs/issue/SKY-297/implement-an-endpoint-which-marks-a-given-server-as-dead . There's nothing stopping us from triggering an extra sweep manually, after a server is pronounced dead.

Ok that is fine with me then. We should make a note to have the process of marking a server as dead trigger a scan, that addresses my main concern.

Reply:
added a note in linear.